### ISSUE TITLE
Hash Collisions in LpVariable

### ISSUE THREAD

**Issue by amakarowsky** [coin-or/pulp#279]
Details for the issue
--------------------

#### What did you do?
1) Write a user-defined class which contains LpVariable fields.
2) Build instances of that class and save them to a local directory (many packages are appropriate for this, I used shelve).
3) Several days later (separate Python session etc.), I loaded the saved objects and attempted to instantiate lpSums, lpAffineExpressions and constraints between their fields.

For example, let myobj1 and myobj2 both contain LpVariable fields namely .num. The variables were instantiated with names that were different for each object.

obj1 = load_from_cache(myobj1)
obj2 = load_from_cache(myobj2)
my_objects = [obj1, obj2]
my_expr = lpSum([o.num for o in my_objects])    <- Interpreter never passes, this line, and instead gives "max recursion depth exceeded"

Note I also tried various sum() and LpAffineExpression() and lpDot() methods. 

I notice myobj1.num == myobj2.num fails with the same error.

The hash of each variable, i.e. hash(myobj1.num) and hash(myobj2.num), are the same.

#### What did you expect to see?
I expect the various LP functions to work as documented and not create 'recursion depth exceeded' errors, especially when the variables themselves have different names.

#### What did you see instead?
Pulp functions appear to call themselves recursively when hash of different variables are the same, leading to an error.

This thread appears to highlight a similar error when constructing between different threads: https://github.com/coin-or/pulp/pull/247

This stack overflow thread may describe something similar: https://stackoverflow.com/questions/50030498/recursion-error-using-pulp

Useful extra information
-------------------------

The info below often helps, please fill it out if you're able to. :)

#### What operating system are you using?

- [ ] Windows: ( _version:_ ___ )
- [X] Linux: ( _distro:_ Ubuntu  )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### I'm using python version:

- [ ] 2.7
- [ ] 3.4
- [ ] 3.5
- [X] 3.6
- [ ] Other: ___

#### I installed PuLP via:

- [X] pypi (pip install pulp)
- [ ] github
- [ ] Other: ___

#### Did you also

- [ ] Tried out the latest github version: https://github.com/coin-or/pulp
- [X] Searched for an existing similar issue: https://github.com/coin-or/pulp/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by amakarowsky** [comment]
Code for reproducing this error is given below:

import pulp as pl
import shelve

v1 = pl.LpVariable("my_var_a_" + str(i), 0, 1, cat = "Integer")
#Save the variable to a file.
save = shelve.open("v1")
save["var"] = v1
save.close()
#Every second instantiation of an LpVariable in a local var collides, hence instantiate a dummy
v1 = pl.LpVariable("my_var_b_" + str(i), 2, 5, cat = "Integer")
#Now a third - note all three variables are different...
v1 = pl.LpVariable("my_var_c_" + str(i), 6, 9, cat = "Integer")
#Load the first variable back.
store = shelve.open("v1")
v2 = store["var"]
store.close()
print(hash(v1))
print(hash(v2))
try:
    v1 == v2
except:
    print("Hash collision")

**Comment by amakarowsky** [comment]
Note the v1 == v2 and print("Hash collision") lines should be indented.

**Comment by pchtsp** [comment]
Hello and thanks a lot for the reproducible example (although you missed an `i=1` somewhere at the beginning I think).

I think `pulp` has never supported saving its objects into files, although maybe @stumitchell can confirm this.

It could be related with the parallel variable generation code since they both crash because of the way the hash for variables is generated.

Maybe we could change that. @stumitchell what about just using the variable name (shouldn't that be unique already?).

Anyway, if you @amakarowsky feel free to come up with a PR that permits variable objects and constraints to be written correctly into files and we would definitively check it.

F.

**Comment by stumitchell** [comment]
Ok the core of the problem here is that LpVariable needs a unique has as the Lp Expressions are stored as dicts where the key is the LpVariable and the value is the coefficient.

In pulp we currently don't require unique variable names, as the user may want to generate variables in a loop etc.

@amakarowsky thanks for your great issue about this it really does get to the root of the problem.

Other api's for instance gurobi get around this by having an explicit tie between the variable and the model.

We mainly have just told people not to try and generate in parallel and don't persist lpVariables.

I think the best solution would be to have a better error when we have a hash collision. Letting the user know what probably caused the issue, and how to avoid.

**Comment by amakarowsky** [comment]
Thank you both @pchtsp and @stumitchell for looking at this. I am a big fan of PuLP and really appreciate your time.

The example above passes for me if I change the way self.hash is initialised within the __init__() method of LpElement. In particular, to fix the case above, I replaced:

self.hash = id(self)
with
self.hash = id(self) + id(name)

It does however support persistence (and potentially parallel construction) as long as users are explicit and intentional in a build-now, solve-later programming approach. I often verify the integrity of a time-consuming problem during the morning at work before I schedule a set of jobs to start at 6PM. Note also that this solution does not force users to create unique variable names. 

To me, requiring everything to be built at run time appears an unnecessary restriction on a great package :+1: 

It is worth noting that if I set the names of the variable equal to one another (i.e. my_var_c -> my_var_a), I still receive a collision. This is arguably desirable: PuLP needs some way of distinguishing two variables, and an error message (if an unwieldy one) would alert the user to a flaw in their build-now, solve-later approach.
__
Going somewhat tinfoil hat, it is likely feasible to generate two memory locations and names for which id(location1) + id(name1) == id(location2) + id(name2), which would produce a collision. I have thought of things like self.hash = int(str(id(self))+ str(id(name))) but these approaches have limits. Truly avoiding such behaviour probably requires a larger refactor that avoids the use of hash values entirely to distinguish LpElements.

**Comment by amakarowsky** [comment]
I'm a little new to PR requests etc. through Github @pchtsp . Happy to have a you if you think the approach above seems sensible however.

**Comment by pchtsp** [comment]
I've been thinking about this (sorry to respond this late). Couldn't have a way to store the model as a json and then recreate the object from it? There's not too much data and I think this would be feasible. I feel a lot more comfortable dealing with jsons to serialize and de-serialize an object (for once, it makes it compatible with other systems). That same logic could be then used to store the pickle (a dictionary of data). What so you think @stumitchell ?

**Comment by amakarowsky** [comment]
Hi @pchtsp , all good for the delay, thank you for thinking further about the issue. I like the json idea, especially if should be widely compatible with other systems

**Comment by pchtsp** [comment]
@amakarowsky if you care to try, there is a branch that has a first json export functionality:
https://github.com/coin-or/pulp/tree/export_pb

check new tests test_pulp_210, test_pulp_211 and test_pulp_220 for an example on how it works. There are "to_dict", "to_json", "from_dict", "from_json" methods on the problem object.

There are some likely changes on the structure (some dictionaries will become lists), although it won't change too much.

Tell me it it works with your problem, since I have not tested it with big problems, only toy ones.

**Comment by amakarowsky** [comment]
Hi @pchtsp, I just had a look at the feature branch.

Thank you immensely for putting it together - I think the desired functionality will enable clean persistence workflows with PuLP that I have been using for a while now.

Is there any chance of incorporating a call to "from_json" within one of the tests (perhaps 211?) I'm somewhat unfamiliar with persisting through .json so a rebuilding example would be really useful for persons like me that may be more familiar with pickle or shelve.

Is there any chance of from_json and to_json methods for LpVariable? For context, I usually solve hundreds of slightly different problems using a saved set of LpVariable objects. I don't typically re-load a single problem object more than once. Persisting individual variables is definitely possible with this arrangement, they each require a dummy problem with a junk name & objective (seems awkward perhaps?).

Great to hear for you once again and I really appreciate the work you have put into this problem.

**Comment by pchtsp** [comment]
Hey, 

I'll try to get a new commit towards the end of the week. I will:

1. slightly change the way the constraints and variables are stored.
2. add from_json examples.

Regarding variables, they already have a "to_dict" and "from_dict" method. I guess you want all variables in one single file, right? maybe if you create a dictionary of variables and export that to json, then you would have to import that again.
I think the best way to explain that (and maybe the "from_json" too) would be to document the use of this functionality in the docs. I'll see if I can come up with that too soon.

If you have a case study in mind, feel free to post in here in detail to use it as inspiration for the example in the docs.

**Comment by pchtsp** [comment]
So, I've just pushed 1 and 2 to the repository. Feel free to test it. If you're feeling generous, you can contribute with writing a small guide on how to import and export a problem or a variable : )

**Comment by amakarowsky** [comment]
Hi @pchtsp  - great work on the development again, I really appreciate it. 

Happy to detail a case study I have been working on at work (at a high level) as an example. I am also happy to make a small guide on importing and exporting.

I have a few deadlines coming up at work at present but I am hoping to have these contributions ready to review by the end of the week. Alex

**Comment by pchtsp** [comment]
The documentation for importing and exporting an LpProblem are described here:

https://coin-or.github.io/pulp/guides/how_to_export_models.html

I'm closing the issue. If you see have any problems with this way of saving a problem, just make a new issue.

**Comment by amakarowsky** [comment]
Wonderful stuff @pchtsp !
